crates/bls-crypto/src/bls/batch.rs

```diff
@@ -0,0 +1,86 @@
+use super::{PublicKey, Signature};
+
+// use algebra::{bls12_377::G1Projective, Field, PrimeField, ToBytes};
+use ark_bls12_377::{Fr, G1Projective};
+use ark_ff::{Field, PrimeField, ToBytes};
+
+use blake2s_simd::{Params, State};
+
+use crate::{BLSError, HashToCurve};
+
+#[derive(Default)]
+pub struct Batch {
+    entries: Vec<(PublicKey, Signature)>,
+    message: Vec<u8>,
+    extra_data: Vec<u8>,
+    size: usize,
+}
+
+impl Batch {
+    /// Constructs a new strict batch verifier context for a given message.
+    pub fn new(message: &[u8], extra_data: &[u8]) -> Batch {
+        Batch {
+            entries: Vec::<(PublicKey, Signature)>::new(),
+            message: message.to_vec(),
+            extra_data: extra_data.to_vec(),
+            size: 0,
+        }
+    }
+
+    pub fn add(&mut self, public_key: &PublicKey, signature: &Signature) {
+        self.entries
+            .push((public_key.to_owned(), signature.to_owned()));
```

Comment:
Same as in another place, maybe best to let the caller know they need to pass owned versions